drivers/usb/udc/udc_renesas_ra.c

```diff
@@ -0,0 +1,756 @@
+/*
+ * Copyright (c) 2024 Renesas Electronics Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "udc_common.h"
+
+#include <soc.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/clock_control/renesas_ra_cgc.h>
+#include <zephyr/drivers/usb/udc.h>
+#include "r_usb_device.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(udc_renesas_ra, CONFIG_UDC_DRIVER_LOG_LEVEL);
+
+struct udc_renesas_ra_config {
+	const struct pinctrl_dev_config *pcfg;
+	size_t num_of_eps;
+	struct udc_ep_config *ep_cfg_in;
+	struct udc_ep_config *ep_cfg_out;
+	void (*make_thread)(const struct device *dev);
+	int speed_idx;
+};
+
+struct udc_renesas_ra_data {
+	struct k_thread thread_data;
+	struct st_usbd_instance_ctrl udc;
+	struct st_usbd_cfg udc_cfg;
+};
+
+enum udc_renesas_ra_event_type {
+	/* An event generated by the HAL driver */
+	UDC_RENESAS_RA_EVT_HAL,
+	/* Shim driver event to trigger next transfer */
+	UDC_RENESAS_RA_EVT_XFER,
+	/* Let controller perform status stage */
+	UDC_RENESAS_RA_EVT_STATUS,
+};
+
+struct udc_renesas_ra_evt {
+	enum udc_renesas_ra_event_type type;
+	usbd_event_t hal_evt;
+	uint8_t ep;
+};
+
+K_MSGQ_DEFINE(drv_msgq, sizeof(struct udc_renesas_ra_evt), CONFIG_UDC_RENESAS_RA_MAX_QMESSAGES,
+	      sizeof(uint32_t));
+
+extern void usb_device_isr(void);
+
+static void udc_event_handler(usbd_callback_arg_t *p_args)
+{
+	const struct device *dev = p_args->p_context;
+	struct udc_renesas_ra_evt evt;
+
+	switch (p_args->event.event_id) {
+	case USBD_EVENT_BUS_RESET:
+		udc_submit_event(dev, UDC_EVT_RESET, 0);
+		break;
+
+	case USBD_EVENT_VBUS_RDY:
+		udc_submit_event(dev, UDC_EVT_VBUS_READY, 0);
+		break;
+
+	case USBD_EVENT_VBUS_REMOVED:
+		udc_submit_event(dev, UDC_EVT_VBUS_REMOVED, 0);
+		break;
+
+	case USBD_EVENT_SUSPEND:
+		udc_submit_event(dev, UDC_EVT_SUSPEND, 0);
+		break;
+
+	case USBD_EVENT_RESUME:
+		udc_submit_event(dev, UDC_EVT_RESUME, 0);
+		break;
+
+	case USBD_EVENT_SOF:
+		udc_submit_event(dev, UDC_EVT_SOF, 0);
+		break;
+
+	default:
+		evt.type = UDC_RENESAS_RA_EVT_HAL;
+		evt.hal_evt = p_args->event;
+		k_msgq_put(&drv_msgq, &evt, K_NO_WAIT);
+		break;
+	}
+}
+
+static void usb_interrupt_handler(void *arg)
+{
+	ARG_UNUSED(arg);
+	usb_device_isr();
+}
+
+static void udc_event_xfer_next(const struct device *dev, const uint8_t ep)
+{
+	struct udc_renesas_ra_data *data = udc_get_private(dev);
+	struct net_buf *buf;
+
+	if (udc_ep_is_busy(dev, ep)) {
+		return;
+	}
+
+	buf = udc_buf_peek(dev, ep);
+	if (buf != NULL) {
+		int err;
+
+		if (USB_EP_DIR_IS_IN(ep)) {
```

Comment:
You need to check udc_ep_buf_has_zlp(buf), see how it is used in existing drivers.

Reply:
I fixed it